[FLINK-38825][streaming] Add ordered async batch support for AsyncBatchWaitOperator

featzhang opened the pull request:
## What is the purpose of the change
This PR adds ordered semantics to the async batch operator, building on top of:
PR1: Size-based batching (AsyncBatchWaitOperator)
PR2: Time-based batching (timeout flush)

For AI inference workloads and other batch-oriented async operations, users may require that output records maintain the same order as input records, even when async batch invocations complete out of order. This is essential for downstream processing that depends on ordering guarantees.

## Brief change log
- Add OrderedAsyncBatchWaitOperator - maintains ordering by assigning monotonic sequence numbers to each batch and emitting results strictly in sequence order
- Add OrderedAsyncBatchWaitOperatorFactory - creates ordered operator instances
- Add AsyncDataStream.orderedWaitBatch() API methods as entry points
- Add comprehensive unit tests in OrderedAsyncBatchWaitOperatorTest

## Verifying this change
This change added tests and can be verified as follows:
- OrderedAsyncBatchWaitOperatorTest#testStrictOrderingGuarantee - verifies batches completing out-of-order still produce ordered output
- OrderedAsyncBatchWaitOperatorTest#testBatchAndTimeoutTriggerWithOrdering - verifies ordering preserved across size and time triggered batches
- OrderedAsyncBatchWaitOperatorTest#testExceptionPropagation - verifies exceptions are propagated correctly
- OrderedAsyncBatchWaitOperatorTest#testDelayedAsyncCompletion - verifies delayed completions are handled correctly
- OrderedAsyncBatchWaitOperatorTest#testManyBatchesOrdering - verifies sequence number handling for many batches

## Does this pull request potentially affect one of the following parts
- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with @Public(Evolving): yes (new API methods)
- The serializers: no
- The runtime per-record code paths (performance sensitive): no
- Anything that affects deployment or recovery: no
- The S3 file system connector: no

## Documentation
- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? JavaDocs

---

## PR Series for FLINK-38825

JIRA: [FLINK-38825](https://issues.apache.org/jira/browse/FLINK-38825) - Introduce an AI-friendly Async Batch Operator for high-latency inference workloads

This feature is implemented incrementally through the following PR series:

| # | PR | Title | Description | Module |
|---|---|---|---|---|
| 1 | [#27355](https://github.com/apache/flink/pull/27355) | Introduce AsyncBatchFunction and AsyncBatchWaitOperator | Core API and runtime operator with unordered semantics and size-based batch triggering | `flink-streaming-java` |
| 2 | [#27356](https://github.com/apache/flink/pull/27356) | Add time-based batch triggering | Timeout-based flush to trigger incomplete batches after a configurable duration | `flink-streaming-java` |
| **3** | [#27357](https://github.com/apache/flink/pull/27357) | Add ordered async batch support | Ordered output semantics via OrderedAsyncBatchWaitOperator with sequence numbers | **`flink-streaming-java`** |
| 4 | [#27358](https://github.com/apache/flink/pull/27358) | Add inference-oriented metrics | Batch size/latency histograms, async call duration, inflight batches, failure counters | `flink-streaming-java` |
| 5 | [#27359](https://github.com/apache/flink/pull/27359) | Implement retry and timeout strategies | Fixed-delay/exponential-backoff retry, fail-on-timeout/allow-partial timeout policies | `flink-streaming-java` |
| 6 | [#27360](https://github.com/apache/flink/pull/27360) | Add SQL/Table API integration | AsyncBatchLookupFunction, AsyncBatchLookupFunctionProvider, lookup join runner | `flink-table` |
| 7 | [#27361](https://github.com/apache/flink/pull/27361) | Add Python DataStream API integration | Python AsyncBatchFunction, async_invoke_batch() API, AsyncDataStream batch methods | `flink-python` |

> **Note**: Each PR builds incrementally on the previous ones. PRs should be reviewed and merged in order.

Commit:
[FLINK-38825] Introduce AsyncBatchFunction and AsyncBatchWaitOperator

Commit:
[FLINK-38825] Add time-based batch triggering for AsyncBatchWaitOperator

Commit:
[FLINK-38825] Add ordered async batch support for AsyncBatchWaitOperator

xishuaidelin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AsyncDataStream.java:
```diff
@@ -319,4 +323,159 @@ public static <IN, OUT> SingleOutputStreamOperator<OUT> orderedWaitWithRetry(
+            DataStream<IN> in,
+            AsyncBatchFunction<IN, OUT> func,
+            int maxBatchSize,
+            Duration maxWaitTime) {
```
Nit: There is an inconsistency in how time intervals are represented in this signature compared to similar methods.
`unorderedWaitBatch(
            DataStream<IN> in,
            AsyncBatchFunction<IN, OUT> func,
            int maxBatchSize,
            long batchTimeoutMs)`

featzhang committed:
[FLINK-38825][streaming] Refactor OrderedAsyncBatchWaitOperator to extend AsyncBatchWaitOperator

- Make OrderedAsyncBatchWaitOperator extend AsyncBatchWaitOperator to eliminate
  duplicated logic (buffer management, timer handling, in-flight tracking)
- Replace TreeMap + sequence number ordering with a simpler FIFO ArrayDeque of
  pre-allocated ResultSlots, which is less complex and equally correct
- Align unorderedWaitBatch() timeout parameter signature with other async methods:
  use (long timeout, TimeUnit timeUnit) instead of (long batchTimeoutMs)
- Fix flaky testStrictOrderingGuarantee: replace Thread.sleep polling with
  CountDownLatch for deterministic synchronization

featzhang committed:
[FLINK-38825][streaming] Fix test failures in AsyncBatchWaitOperator

- Fix NullPointerException by calling super(parameters) in AsyncBatchWaitOperator
  constructor to properly initialize processingTimeService
- Fix UnsupportedOperationException in testExceptionPropagation by setting
  expectedExternalFailureCause on MockEnvironment before running the test
- Fix testExceptionPropagation timeout in OrderedAsyncBatchWaitOperator by
  completing the ResultSlot and draining the result queue in completeExceptionally
  to avoid infinite loop in endInput()

featzhang commented:
## Review Response Summary

Thank you @xishuaidelin for the excellent feedback! I've analyzed all 4 comments:

### Comments to Address:
1. **Code duplication** - Extend AsyncBatchWaitOperator instead
2. **Complex TreeMap ordering** - Use simple FIFO queue pattern
3. **API inconsistency** - Parameter naming mismatch
4. **Flaky Thread.sleep()** - Replace with CountDownLatch

### Major Refactoring:

**1. Extend Base Operator:**
```java
public class OrderedAsyncBatchWaitOperator<IN, OUT> 
        extends AsyncBatchWaitOperator<IN, OUT> {
    
    private final Queue<CompletedBatch<OUT>> completedBatches = new ArrayDeque<>();
    private long nextExpectedSequence = 0;
    
    @Override
    protected void invokeBatchAsync(List<IN> batch) {
        final long sequence = nextBatchSequence++;
        function.asyncInvoke(batch, new ResultFuture<OUT>() {
            public void complete(Collection<OUT> results) {
                completedBatches.add(new CompletedBatch<>(sequence, results));
                emitCompletedBatches();
            }
        });
    }
}
```

**2. Simple FIFO Queue:**
```java
private void emitCompletedBatches() {
    while (!completedBatches.isEmpty()) {
        CompletedBatch<OUT> head = completedBatches.peek();
        if (head.sequence != nextExpectedSequence) break;
        completedBatches.poll();
        emit(head.results);
        nextExpectedSequence++;
    }
}
```

**3. Robust Testing:**
```java
CountDownLatch latch = new CountDownLatch(expectedCount);
// ... async operations
assertTrue(latch.await(10, TimeUnit.SECONDS));
```

This refactoring eliminates ~70% of code duplication and simplifies ordering logic significantly.

cc @featzhang
